### PR TITLE
sql: optimize regexp_replace for static regexes

### LINKED ISSUE #22172: Optimize `regexp_replace` for static regexes
### Feature request

We should add an optimized implementation of `rexexp_replace` that compiles the regex only once if it is a static string. We already do this for `regexp_match` (just grep for "regexp_match_static").

### PR DESCRIPTION
**Pull request by sdht0**
Mirroring the `regexp_match` sql function, this PR pre-processes static regexes for `regexp_replace`.

Fixes #22172 

### Motivation

The `Regex` object is now constructed from their string representation only once when possible, instead of constructing the same object repeatedly across multiple rows.

### Tips for reviewer

- The implementation follows the style of `regexp_match`, except we need a binary function here.
- Because of the `g` flag, we need to carry around an extra integer. I created a new proto object `ProtoRegexReplace` to be able to represent this.
- I've extended the tests to ensure the fast and slow paths are taken appropriately. 
- [x] I've tried to guess the correct output for the functions in `impl BinaryFunc`. These should be verified properly though, as I don't yet understand the exact implications of these functions.
- [x] The code for computing `n` is repeated in two places and should be hoisted out to a helper function. What's the best file for that?

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by ggevay** [comment]
The CI test fail in `cargo test` looks unrelated to this PR. I've pinged the Storage team on Slack.

File: src/expr/src/scalar/func.rs [hunk 1]
```diff
@@ -2791,6 +2769,7 @@ impl BinaryFunc {
 
             ParseIdent => ScalarType::Array(Box::new(ScalarType::String)).nullable(in_nullable),
             PrettySql => ScalarType::String.nullable(in_nullable),
+            RegexpReplace { .. } => ScalarType::Array(Box::new(ScalarType::String)).nullable(true),
```
[the remark this answers is not in the document]
Comment (ggevay) [reply src/expr/src/scalar/func.rs:2772]: Also, it shouldn't be an array. It should be exactly the same as for the original, variadic `RegexpReplace`, see [here](https://github.com/MaterializeInc/materialize/blob/f89bb6051898d7c8aa7e5cd39816921b7e8ae86d/src/expr/src/scalar/func.rs#L7871).

This is the cause of the CI test failure in `Fast SQL logic tests`. You can reproduce this locally by:
```
RUST_BACKTRACE=1 bin/sqllogictest -- -v --auto-index-selects test/sqllogictest/redacted.slt
```

Note that the reason why `regex.slt` didn't catch this is that it's not run with `--auto-index-selects` (because [it's in the](https://github.com/MaterializeInc/materialize/blob/77275f5565e001e9abaadea9b041f4c97686d646/test/sqllogictest/mzcompose.py#L516) `tests_without_views` category). If you run `regex.slt` with `--auto-index-selects` locally, then it also happens there. (What `--auto-index-selects` does is that for each SELECT in the slt file, it also creates an indexed view of it, and selects from the indexed view. This is to catch bugs that happen only with a long-running dataflow, and not for a one-off SELECT.) 

But I don't understand why is this happening only with `--auto-index-selects`. We should definitely figure this out.

[the remark this answers is not in the document]
Comment (ggevay) [reply src/expr/src/scalar/func.rs:2772]: I'm also not sure why is `regex.slt` in the `tests_without_views` category. Maybe we should remove it from there. On `main`, it runs fine with `--auto-index-selects`. @MaterializeInc/testing, do you know what are the possible reasons in general why we put slts in `tests_without_views`? I thought it's only when it would fail with `--auto-index-selects` due to legitimate reasons (e.g., different EXPLAIN output for a long-running dataflow). Is it also due to slowness reasons? If `regex.slt` was there only due to slowness reasons, then I'd say running `regex.slt` with `--auto-index-selects` would be actually worth the time. There have been so many tricky regex bugs that testing them as thoroughly as possible is important.

**Review by ggevay — CHANGES_REQUESTED** [review]

File: src/expr/src/scalar/func.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrainer-materialize) [reply src/expr/src/scalar/func.rs:2772]: Unfortunately, I don't have a good understanding of why files are in a certain category. Although, I touched the test file, I only undertook semantic-preserving refactorings.
The comment in the code "renable a fast subset of the following when performance is restored" hints that performance can be a reason for a file to be assigned to `tests_without_views`.

Feel free to change assignments if you are somewhat confident with your change!

**Comment by nrainer-materialize** [comment]
Please trigger a nightly run and a SLT run before merging this PR. Thank you!

File: src/expr/src/scalar/func.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ggevay) [reply src/expr/src/scalar/func.rs:2772]: @sdht0, could you try removing `"test/sqllogictest/regex.slt",` from the `tests_without_views` list? I think it should run fine after fixing the type bug.

[the remark this answers is not in the document]
Comment (sdht0) [reply src/expr/src/scalar/func.rs:2772]: Thanks for the review @ggevay! I will address the comments later today.

[the remark this answers is not in the document]
Comment (ggevay) [reply src/expr/src/scalar/func.rs:2772]: Oh, I think I know why the type bug was getting caught by slts only with `--auto-index-selects`: The check that was failing is based on the `RelationDesc`, and I think for one-off SELECTs the `RelationDesc` is created based on the HIR plan (see `plan_root_query`), which is before the MIR optimizations are run, but for an indexed view we rely on the type that is after MIR optimization. @mgree, @aalexandrov, one more reason why we should add a check that the type of a query after optimization matches the type that was before optimization.

**Comment by sdht0** [comment]
@ggevay sorry for the delay! I had pushed the changes last night, but was investigating the behavior of `--auto-index-selects`.

It seems like even after removing `regex.slt` from `tests_without_views`, the `--auto-index-selects` flag is still needed to detect the earlier bug in `output_type()`. Without the flag, `bin/sqllogictest -- test/sqllogictest/regex.slt` still passes.

I also refactored the code to: (i) rename `n` to `limit`, reflecting the name in `Regex::replacen`; and (ii) create a helper function to avoid duplicated code for parsing the `g` flag.

File: src/expr/src/scalar.rs [hunk 2]
```diff
@@ -1181,6 +1182,38 @@ impl MirScalarExpr {
...
+                            // path. We could switch build_regex to take an iter which would also achieve that.
+                            let (n, flags) = if flags.contains('g') {
+                                let flags = flags.replace('g', "");
+                                (0, Cow::Owned(flags))
+                            } else {
+                                (1, Cow::Borrowed(flags))
+                            };
+                            *e = match func::build_regex(pattern, &flags) {
+                                Ok(regex) => {
+                                    let mut exp = mem::take(exprs);
```
[the remark this answers is not in the document]
Comment (sdht0) [reply src/expr/src/scalar.rs]: Done, thanks!

**Comment by ggevay** [comment]
Thanks, I'll continue the review later today.

> It seems like even after removing `regex.slt` from `tests_without_views`, the `--auto-index-selects` flag is still needed to detect the earlier bug in `output_type()`. Without the flag, `bin/sqllogictest -- test/sqllogictest/regex.slt` still passes.

This is ok: removing `regex.slt` from `tests_without_views` just means that CI will run `regex.slt` with `--auto-index-selects`, see [here](https://github.com/MaterializeInc/materialize/blob/77275f5565e001e9abaadea9b041f4c97686d646/test/sqllogictest/mzcompose.py#L521-L527).

**Comment by ggevay** [comment]
I've started a Nightly and a full SLT:
https://buildkite.com/materialize/nightly/builds/7908
https://buildkite.com/materialize/slt/builds/5939

**Comment by ggevay** [comment]
In the Nightly, the `K8s recovery: envd on failing node` fail is probably unrelated, but the `Output consistency` seems to be failing on something that might be a real bug: https://buildkite.com/materialize/nightly/builds/7908#018fbf02-d46d-45f8-a324-1c945d19e7fc/1000
It seems that errors in regexes or flags are handled differently in the new code compared to `main`. Could you please look into this @sdht0? If you find that only the error msgs changed slightly, that would be ok, but if I'm reading the test output correctly, we somehow change between erroring and not erroring, which seems to be a real problem.

**Comment by nrainer-materialize** [comment]
> In the Nightly, the `K8s recovery: envd on failing node` fail is probably unrelated, but the `Output consistency` seems to be failing on something that might be a real bug: https://buildkite.com/materialize/nightly/builds/7908#018fbf02-d46d-45f8-a324-1c945d19e7fc/1000 It seems that errors in regexes or flags are handled differently in the new code compared to `main`. Could you please look into this @sdht0? If you find that only the error msgs changed slightly, that would be ok. But if I'm reading the test output correctly, we somehow change between erroring and not erroring, which seems to be a real problem.

Yes, the K8s node recovery test just timed out. I triggered a retry; though, I am sure the failure is unrelated to the changes.
The output version consistency check indeed uncovered a relevant issue! That's something we definitively need to look into.

**Comment by sdht0** [comment]
Right so from a first look, I don't understand how these tests were passing before:
- In `regexp_replace(varchar_8_null, 'ab(', text_val_5, 'i')`, `ab(` indeed does have unbalanced paranthesis.
- And is `regexp_replace(text_null, '[ab]', char_6_val_w_spaces, 'invalid_value_123')` a special syntax for this function? What is the 3th or 4th parameter doing here?

**Comment by ggevay** [comment]
These tests are verifying that the system correctly reacts to error cases, e.g., when the user gives an invalid regex or flag. So, the test is expecting an error from running these queries.

**Comment by nrainer-materialize** [comment]
> Right so from a first look, I don't understand how these tests were passing before:
> 
>     * In `regexp_replace(varchar_8_null, 'ab(', text_val_5, 'i')`, `ab(` indeed does have unbalanced paranthesis.
> 
>     * And is `regexp_replace(text_null, '[ab]', char_6_val_w_spaces, 'invalid_value_123')` using a special syntax for this function? What is the 3th or 4th parameter doing here?

You can find the function definition [here](https://materialize.com/docs/sql/functions/) by searching for `regexp_replace`:
```
regexp_replace(source: str, pattern: str, replacement: str [, flags: str]]) -> str`
```

**Comment by sdht0** [comment]
Maybe I don't understand the output then. For instance:
```
-- Errors:
-- * ValidationErrorType.SUCCESS_MISMATCH: Outcome differs.
--   Value 1 (Dataflow rendering v0.102.0-dev (8027f64b0)): 'QueryFailure' (type: <class 'str'>)
--   Value 2 (Dataflow rendering v0.102.0-dev (02b35cac7)): 'QueryResult' (type: <class 'str'>)
--   Error 1: 'Evaluation error: invalid regular expression flag: n'
--   Query 1: SELECT regexp_replace(text_null, 'ab(', varchar_8_val_1, 'invalid_value_123') FROM t_dfr_1_v0_102_0_dev_horiz;
--   Query 2: SELECT regexp_replace(text_null, 'ab(', varchar_8_val_1, 'invalid_value_123') FROM t_dfr_2_v0_102_0_dev_horiz;
```
Is this indicating that it expected a `QueryResult` but is getting a `QueryFailure`? If so, I'm not sure how the `varchar_8_val_1` is being used here. I am guessing `invalid_value_123` is supposed to map to `replacement` in `regexp_replace()`. Is `varchar_8_val_1` then changing the locale?

The new code seems to be mapping `invalid_value_123` to  `flags` instead, hence the error.

> And is `regexp_replace(text_null, '[ab]', char_6_val_w_spaces, 'invalid_value_123')` using a special syntax for this function? What is the 3th or 4th parameter doing here?

Same for `char_6_val_w_spaces` here.

**Comment by ggevay** [comment]
`8027f64b0` is this PR, and `02b35cac7` is `main`. 

> `Value 1 (Dataflow rendering v0.102.0-dev (8027f64b0)): 'QueryFailure' (type: <class 'str'>)`

This says that the query failed in this PR. 

> --   Value 2 (Dataflow rendering v0.102.0-dev (02b35cac7)): 'QueryResult' (type: <class 'str'>)

This says that the same query succeeded on `main`.

`varchar_8_val_1` should be the replacement, and `'invalid_value_123'` should be the flag on both main and this PR.

**Comment by ggevay** [comment]
Oh, the problem might be an inconsistency in null handling. Maybe `main` returns null when the first input is null even with a wrong regex, but the new code fails the query for a wrong regex even for a null first input. Unfortunately, this might be an unavoidable change with this optimization, so we might have to just adjust the test instead of fixing any code. I have to run to a meeting, will get back to this afterwards.

**Comment by sdht0** [comment]
Thanks for the explanations! I'll look into this more after my interview with Materialize :)

**Comment by sdht0** [comment]
> So, adding a bunch of test ignores might be a shortcut "solution" to both of these problems, but I'm tempted to do the right thing instead (see above at "noting down in a new field in the BinaryFunc that we have an error"),

Yes I prefer the later too, as making the code more complex to satisfy test conditions doesn't seem right.

>  let me know if you'd prefer a call instead of GitHub comments, because this is getting a bit long-winded in writing.

Absolutely. Did you have Zoom in mind? You can send an invite to s3sahu[at]uwaterloo.ca anytime today.

**Comment by ggevay** [comment]
What time would be good for you? For example, any time in the next 60 minutes would be good for me, or tomorrow 10:00 - 10:30 or 11:30 - 12:00 Toronto time.

**Comment by nrainer-materialize** [comment]
> Another option is to just tweak the test to ignore the output inconsistency where one errors and the other succeeds with a null result. How hard would this be, @nrainer-materialize? Maybe we've already run into the same issue for other functions, and we already have some standard way of doing this? For example, maybe the very similar optimization of `regex_match` has the same issue?

Adding an ignore would be rather easy for the observed failure. Happy to support on that if needed but looks as you prefer to pursue another approach, which is perfectly fine for me.

> Here, it seems we are changing whether we look at the 2nd or the 4th argument first, and thus observe different errors in different arguments. The original code just evaluates each argument in order, and picks the first error that happens, while the new code errors out when `build_regex` is checking the flags. I'm not sure why `regexp_match` is not running into the same problem. @nrainer-materialize, is there some ignore for `regexp_match` for differing error msgs?

The failure occurred in the version-consistency framework where no ignores should apply for recent versions. I can check tomorrow if `regexp_match` is specified properly.

Furthermore, I wanted to mention that the output consistency framework has different targets. We have seen failures only in the _version-consistency_ (comparing against a previous version of mz). The framework also tests _dataflow-rendering vs. constant folding_ (internal output consistency) and _Postgres consistency_. I am a bit surprised that the latter did not fail as well but maybe there are some eager ignores in place.

**Comment by ggevay** [comment]
> The framework also tests dataflow-rendering vs. constant folding (internal output consistency) and Postgres consistency. I am a bit surprised that the latter did not fail as well but maybe there are some eager ignores in place.

Hmm, yes, we should look into why "internal output consistency" didn't catch the same things, as it seems very relevant for this PR.

@sdht0, the "dataflow-rendering vs. constant folding (internal output consistency)" is the test that I mentioned on the call, which doesn't compare between different versions, but runs the same query on constant inputs (where your optimization would kick in) vs. non-constant inputs on the same system version.

**Comment by ggevay** [comment]
@nrainer-materialize, could you please tell us a command that runs these tests locally?

**Comment by nrainer-materialize** [comment]
> Hmm, yes, we should look into why "internal output consistency" didn't catch the same things, as it seems very relevant for this PR.

So, would you expect to behave dataflow-rendering and constant folding to behave differently with these changes?

> @nrainer-materialize, could you please tell us a command that runs these tests locally?
To run DFR vs. CTF, you can just trigger `bin/output-consistency-test` if you have a running mz instance. See also: `bin/output-consistency-test --help`
You can also use `mzcompose`, which will spin up the db for you:
```
bin/mzcompose --find output-consistency down -v && bin/mzcompose --find output-consistency run default
```

(or with `postgres-consistency` or `version-consistency`)

**Comment by ggevay** [comment]
> So, would you expect to behave dataflow-rendering and constant folding to behave differently with these changes?

Yes, the new optimization concerns the situation when at least some of the inputs to `regexp_replace` are constants. The version consistency fails were due to the new optimization firing for constant inputs, vs. the old version executing the old, non-constant code path. So I would expect the "internal output consistency" to find approximately the same problems as the consistency between versions.

**Comment by nrainer-materialize** [comment]
> > So, would you expect to behave dataflow-rendering and constant folding to behave differently with these changes?
> 
> Yes, the new optimization concerns the situation when at least some of the inputs to `regexp_replace` are constants. The version consistency fails were due to the new optimization firing for constant inputs, vs. the old version executing the old, non-constant code path. So I would expect the "internal output consistency" to find approximately the same problems as the version-consistency.

I have had a look at the function definitions, ignore entries, and log files.
* The log file of internal-output-consistency does not reveal whether it included `regexp_replace` or `regexp_match` in its tests. Successful queries are not printed and functions are chosen randomly. I will extend the framework to print some more statistics.
* The misc/python/materialize/output_consistency/ignore_filter/internal_output_inconsistency_ignore_filter.py do not contain any entry for regex-related stuff.
* `regexp_match` is specified as function (`misc/python/materialize/output_consistency/input_data/operations/string_operations_provider.py`); however, due to its array return type, it will get used only very seldomly.

**Comment by sdht0** [comment]
Alright, so I finally managed to get this working. Proto <-> Rust types mapping are not trivial for sure.

Errors are now saved in the BinaryFunc struct as a `Result` and returned later during evaluation.

**Comment by ggevay** [comment]
Thanks! I've kicked off standard CI and [Nightly](https://buildkite.com/materialize/nightly/builds/7921). I'll look through to code tomorrow.

**Comment by ggevay** [comment]
Output consistency went through!

**Comment by sdht0** [comment]
None of the tests failures look relevant, do they? Note sure about the `Issue is referenced in comment but already closed: #19097` error in particular.

Code-wise, one design question is about the layout of `regex: Result<(Regex, usize), EvalError>`. This is the most straightforward implementation I got working. Perhaps there is a more generic design we can come up with that can be applied to other refactors in the future.

**Comment by nrainer-materialize** [comment]
> None of the tests failures look relevant, do they? Note sure about the `Issue is referenced in comment but already closed: #19097` error in particular.

I can confirm that failing tests in the nightly pipeline are unrelated.

As for the failing linting step, I think you will just need to run `cargo fmt` to fix it.

File: test/sqllogictest/regex.slt [hunk 3]
```diff
@@ -547,3 +576,109 @@ query T
...
+SELECT s, regexp_replace(s, regex, replacement, 'ig') FROM r;
+----
+FOOBARBAZ
+FOOXX
+foobarbaz
+fooXX
+FOOZARZAZ
+FOOZARZAZ
+foobarbaz
+foo>ar<>az<
```
Comment (ggevay) [inline test/sqllogictest/regex.slt:684]: Would be great to add some test cases that show the error path in action. (with constant input, non-constant input, with `query error` and EXPLAIN)

You can make slt expect an error instead of some output rows by writing
```
query error xxxxxxxxxx
```
at the top of the test case (instead of something like `query TT`).

(And link https://github.com/MaterializeInc/materialize/issues/27355 in a comment, because the error msg will look weird in the EXPLAIN.)

**Comment by sdht0** [comment]
> As for the failing linting step, I think you will just need to run `cargo fmt` to fix it.

Ah right I had done it, but it some of the code that I have not touched is also getting formatted. Should I commit this as well and push?
```diff
diff --git a/src/expr/src/scalar/func.rs b/src/expr/src/scalar/func.rs
index 80ed8769cd..890ae441cf 100644
--- a/src/expr/src/scalar/func.rs
+++ b/src/expr/src/scalar/func.rs
@@ -2231,8 +2231,12 @@ pub enum BinaryFunc {
     Gt,
     Gte,
     LikeEscape,
-    IsLikeMatch { case_insensitive: bool },
-    IsRegexpMatch { case_insensitive: bool },
+    IsLikeMatch {
+        case_insensitive: bool,
+    },
+    IsRegexpMatch {
+        case_insensitive: bool,
+    },
     ToCharTimestamp,
     ToCharTimestampTz,
     DateBinTimestamp,
@@ -2256,9 +2260,15 @@ pub enum BinaryFunc {
     TimezoneIntervalTime,
     TimezoneOffset,
     TextConcat,
-    JsonbGetInt64 { stringify: bool },
-    JsonbGetString { stringify: bool },
-    JsonbGetPath { stringify: bool },
+    JsonbGetInt64 {
+        stringify: bool,
+    },
+    JsonbGetString {
+        stringify: bool,
+    },
+    JsonbGetPath {
+        stringify: bool,
+    },
     JsonbContainsString,
     JsonbConcat,
     JsonbContainsJsonb,
@@ -2278,7 +2288,9 @@ pub enum BinaryFunc {
     TrimLeading,
     TrimTrailing,
     EncodedBytesCharLength,
-    ListLengthMax { max_layer: usize },
+    ListLengthMax {
+        max_layer: usize,
+    },
     ArrayContains,
     ArrayLength,
     ArrayLower,
@@ -2300,8 +2312,13 @@ pub enum BinaryFunc {
     GetByte,
     ConstantTimeEqBytes,
     ConstantTimeEqString,
-    RangeContainsElem { elem_type: ScalarType, rev: bool },
-    RangeContainsRange { rev: bool },
+    RangeContainsElem {
+        elem_type: ScalarType,
+        rev: bool,
+    },
+    RangeContainsRange {
+        rev: bool,
+    },
     RangeOverlaps,
     RangeAfter,
     RangeBefore,

```

**Comment by nrainer-materialize** [comment]
Yes, I think that will be necessary to make the pipeline green.

**Comment by sdht0** [comment]
Ah I see you already fixed the actual issue in #27343.

**Comment by ggevay** [comment]
> Ah right I had done it, but some of the code that I have not touched is also getting formatted. Should I commit this as well and push?

Yes, please. I'm not sure why the existing code is getting reformatted, but I usually just follow the rule
"`cargo fmt` is always right", unless some really bad formatting occurs :)

**Comment by sdht0** [comment]
Done.

> I'm not sure why the existing code is getting reformatted,

I assume it's because the code was merged without running `cargo fmt`? Or the formatting rules might have changed since the last time this file was touched.

**Comment by ggevay** [comment]
> I assume it's because the code was merged without running cargo fmt? Or the formatting rules might have changed since the last time this file was touched.

Well, I would think that both of these scenarios should be prevented by CI, because CI always runs `cargo fmt check`, and we can merge a PR only if CI is green (except Nightly).

**Review by ggevay — COMMENTED** [review]
Ok, this is very close to merging, thanks for iterating on it! I have just some minor comments.

File: test/sqllogictest/regex.slt (hunk 3 above)
Comment (sdht0) [reply test/sqllogictest/regex.slt:684]: For the `EXPLAIN` tests, I'm getting:
`FAIL: error: running file regexr.slt: Couldn't split "EXPLAIN\nSELECT regexp_replace('foobarbaz', 'b(..', 'X', 'ig');\n---\nExplained Query (fast path):\n  Error \"invalid regular expression: regex parse error:\n    b(..\n     ^\nerror: unclosed group\n\nTarget cluster: quickstart\n\nEOF\n\nquery error invalid regular expression flag: x\nSELECT regexp_replace('foobarbaz', 'b(..', 'X', 'igx');\n\nstatement ok\nCREATE TABLE e(s string, regex string, replacement string);\n\nstatement ok\nINSERT INTO e VALUES ('foobarbaz', 'b(..', 'X');\n\nquery error invalid regular expression: regex parse error:\nSELECT s, regexp_replace(s, regex, replacement, 'ig') FROM e;\n\nquery error invalid regular expression flag: x\nSELECT regexp_replace(s, regex, replacement, 'igx') FROM e;\n" at \r?\n----`

`--rewrite-results` similarly fails. How can I write the expected output?

File: src/expr/src/scalar/func.rs [hunk 4]
```diff
@@ -3003,7 +2986,7 @@ impl BinaryFunc {
             | ListLengthMax { .. }
             | ArrayLength
             | ArrayLower
-            | ArrayUpper => true,
+            | ArrayUpper { .. } => true,
```
[the remark this answers is not in the document]
Comment (sdht0) [reply src/expr/src/scalar/func.rs:3006]: Stray refactor, sorry.

File: src/expr/src/scalar.rs [hunk 5]
```diff
@@ -1181,6 +1182,25 @@ impl MirScalarExpr {
...
+                        {
+                            let pattern = exprs[1].as_literal_str().unwrap();
+                            let flags = exprs
+                                .get(3)
+                                .map_or("", |expr| expr.as_literal_str().unwrap());
+                            let (limit, flags) = regexp_replace_parse_flags(flags);
+
+                            let regex = match func::build_regex(pattern, &flags) {
+                                Ok(regex) => Ok((regex, limit)),
+                                Err(err) => Err(err),
```
[the remark this answers is not in the document]
Comment (sdht0) [reply src/expr/src/scalar.rs:1197]: Done.

File: test/sqllogictest/regex.slt (hunk 3 above)
Comment (sdht0) [reply test/sqllogictest/regex.slt:684]: @ggevay ^ in case you missed the comment

**Comment by sdht0** [comment]
> Well, I would think that both of these scenarios should be prevented by CI, because CI always runs `cargo fmt check`, and we can merge a PR only if CI is green (except Nightly).

Right. The reason is more interesting. It seems the presence of the new `RegexpReplace` variant is affecting the formatting of the rest of the variants. I don't think it's ideal to touch otherwise unrelated code, but that's what the `cargo fmt` wants to do.

File: test/sqllogictest/regex.slt (hunk 3 above)
Comment (ggevay) [reply test/sqllogictest/regex.slt:684]: Hmm, I don't see the problem from this error msg either. Could you please point me to a branch where you have this? I'd like to debug it.

Comment (sdht0) [reply test/sqllogictest/regex.slt:684]: Here you go: https://github.com/sdht0/materialize/tree/sql-regexp-replace

Comment (ggevay) [reply test/sqllogictest/regex.slt:684]: This branch seems to be the current state of the PR, where `regex.slt` is passing. I meant a branch where the above error is happening.

Comment (sdht0) [reply test/sqllogictest/regex.slt:684]: Right too early in the morning :sweat_smile:
https://github.com/sdht0/materialize/tree/sql-regexp-replace-2

Comment (sdht0) [reply test/sqllogictest/regex.slt:684]: New file: `test/sqllogictest/regexr.slt`

Comment (ggevay) [reply test/sqllogictest/regex.slt:684]: Thanks! :)

Comment (ggevay) [reply test/sqllogictest/regex.slt:684]: One `-` missing on line 16.

Comment (sdht0) [reply test/sqllogictest/regex.slt:686]: Obvious in hindsight. Sorry, I was looking in the wrong place.

I've pushed the error path testing code all tests pass: `PASS: warning=1 success=73 total=74`. This warning is present even in `main` fyi.

Comment (ggevay) [reply test/sqllogictest/regex.slt:686]: Ok, thanks!

**Review by ggevay — APPROVED** [review]
Thanks for the contribution! Merging (as soon as CI finishes)!

**Comment by sdht0** [comment]
@ggevay @nrainer-materialize thanks a lot for your time and help! You have been great.
